### PR TITLE
docs(adr): add development-warnings adr

### PR DESCRIPTION
**Pull request by joshblack**
ADR for Development Warnings in `@primer/react`

[Rendered markdown](https://github.com/primer/react/blob/7e423651a13836a1ecebd0198c2d939fa37f027b/contributor-docs/adrs/adr-012-development-warnings.md)

[Pull Request](https://github.com/primer/react/pull/2901)

### CODE REVIEW

**Review by mperrotti — APPROVED** [review]
I like this direction! Once this merges, could we create an issue to implement the `warning()` helper?

File: contributor-docs/adrs/adr-XXX-development-warnings.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+> Contributors may wrap hooks within a `__DEV__` block even though hooks are not
+> meant to be called conditionally. This is because the `__DEV__` check can be
+> considered constant in that it will always be true or false for the
+> environment (development or production).
+
+When a contributor would like to communicate a warning to a developer, they
+should use the `warning()` helper.
+
+```ts
+warning(condition, 'This is the message that is logged when condition is false-y')
```
Comment (colebemis) [inline contributor-docs/adrs/adr-XXX-development-warnings.md:47]: I would have guessed that the message would display if the condition is truthy. I read that code as "warn if _condition_"

If this is a convention, I'm fine to keep it. But I figured I'd mention my initial confusion.

If the name of the helper was something like `assert()`, I think falsey would make more sense.

Comment (joshblack) [reply contributor-docs/adrs/adr-XXX-development-warnings.md:47]: Totally makes sense! Happy to flip it if that's more intuitive for folks 👍 

I think the convention stems from [`invariant()`](https://github.com/facebook/fbjs/blob/51e174bead2c0bb6bb9c62d7cd2d63338bcef2d6/packages/fbjs/src/__forks__/invariant.js#L30) ultimately, and things like `assert()`, that frame things from the perspective of "this condition must be met". When it's not met, throw an error, log a warning, etc

To contextualize the two, here is the warning from `Heading` in the two styles:

```tsx
// Warn if condition is true
warning(
  innerRef.current && !(innerRef.current instanceof HTMLHeadingElement),
  'This Heading component should be an instanceof of h1-h6'
);
```

```tsx
// Warn if condition is false
warning(
  innerRef.current && innerRef.current instanceof HTMLHeadingElement,
  'This Heading component should be an instanceof of h1-h6'
);
```

**Comment by joshblack** [comment]
Kind of a break-out from: https://github.com/primer/react/pull/2928#discussion_r1117264832

Would we want this to also address unrecoverable situations? Or keep it scoped to development warnings?

By unrecoverable, I think of things like FB's `invariant()` helper that allows you to state conditions that must be met in order to proceed. If they aren't met, an error is thrown with a custom message.

I feel like TS covers a lot of this but that there sometimes is validation that needs to happen if the type system cannot cover a situation.

File: contributor-docs/adrs/adr-XXX-development-warnings.md (hunk 1 above)
Comment (broccolinisoup) [reply contributor-docs/adrs/adr-XXX-development-warnings.md:47]: Thanks Josh! Warning when the condition truthy feels more intuitive to me. That said, it is totally cool to me if we want follow the convention that it stems from.

**Review by broccolinisoup — APPROVED** [review]

**Comment by lesliecdubs** [comment]
@joshblack so far it seems most folks are in alignment here, but if you'd like me to assign a decision-maker on this one give me a shout.

**Pull request merged**